File: lib/src/main/java/com/sha/formvalidator/validator/DateValidator.java
package com.sha.formvalidator.validator;

import android.annotation.SuppressLint;
import android.text.TextUtils;
import android.widget.EditText;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;


public class DateValidator extends Validator {
    private String[] formats;

    public DateValidator(String errorMessage, String _format) {
        super(errorMessage);

        formats = !TextUtils.isEmpty(_format) ?
                _format.split(";") :
                new String[]{"DefaultDate", "DefaultTime", "DefaultDateTime"};
    }

    @SuppressLint("SimpleDateFormat")
    @Override
    public boolean isValid(EditText et) {
        if (TextUtils.isEmpty(et.getText()))
            return true;
        String value = et.getText().toString();

        for (String _format : formats) {
            DateFormat format;

            switch (_format){
                case "DefaultDate":
                    format = SimpleDateFormat.getDateInstance();

                    break;

                case "DefaultTime":
                    format = SimpleDateFormat.getTimeInstance();

                    break;

                case "DefaultDateTime":
                    format = SimpleDateFormat.getDateTimeInstance();

                    break;

                default:
                    format = new SimpleDateFormat(_format);
                    break;

            }

            Date date;

            try {
                date = format.parse(value);
            } catch (ParseException e) {
                return false;
            }

            if (date != null) return true;
        }

        return false;
    }

}

File: lib/src/main/java/com/sha/formvalidator/validator/DateValidator.kt
package com.sha.formvalidator.validator

import android.annotation.SuppressLint
import android.text.TextUtils
import android.widget.EditText

import java.text.DateFormat
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date


class DateValidator(errorMessage: String, _format: String) : Validator(errorMessage) {
    private val formats: Array<String>

    init {

        formats = if (!TextUtils.isEmpty(_format))
            _format.split(";".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
        else
            arrayOf("DefaultDate", "DefaultTime", "DefaultDateTime")
    }

    @SuppressLint("SimpleDateFormat")
    override fun isValid(et: EditText): Boolean {
        if (TextUtils.isEmpty(et.text))
            return true
        val value = et.text.toString()

        for (_format in formats) {
            val format: DateFormat

            when (_format) {
                "DefaultDate" -> format = SimpleDateFormat.getDateInstance()

                "DefaultTime" -> format = SimpleDateFormat.getTimeInstance()

                "DefaultDateTime" -> format = SimpleDateFormat.getDateTimeInstance()

                else -> format = SimpleDateFormat(_format)
            }

            val date: Date?

            try {
                date = format.parse(value)
            } catch (e: ParseException) {
                return false
            }

            if (date != null) return true
        }

        return false
    }

}
